File: Problem-2_Design_Minstack.py
# Time Complexity : O(1) - top, push, pop, getMin.
# Space Complexity : O(n) - n : length of the stack
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : At first I maintained min as a global variable but couldn't handle if the element popped off. Also
# stored 2 values - (value, min) in a node which was as same as using two stack.
#
# Your code here along with comments explaining your approach

class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.curr_min = float('inf')

    # if the value to be pushed is less than the current minimum then, push the current minimum so as to keep track if x gets popped in future
    # and update the current minimum and then push the element.
    def push(self, x: int) -> None:
        if x <= self.curr_min:
            self.stack.append(self.curr_min)
            self.curr_min = x
        self.stack.append(x)

    # If the element popped is same as the current minimum, then current minimum is updated which is the value before it (its the order in
    # which we pushed) and is also popped off.
    def pop(self) -> None:
        if len(self.stack) < 1:
            print("ERROR: Stack Underflow")
            return
        popped_elem = self.stack.pop(-1)
        if popped_elem == self.curr_min:
            self.curr_min = self.stack.pop(-1)

    def top(self) -> int:
        if len(self.stack) < 1:
            print("ERROR: Stack Underflow")
            return
        return self.stack[-1]

    def getMin(self) -> int:
        return self.curr_min

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: Problem 2-Design MinStack.java
// Time Complexity : O(1)
// Space Complexity : If n be the number of elements in the stack then O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach

I am using custom linked list to implement this stack , the next element would be the next element pushed in the stack ,
and the next min would be storing the pointer to the next minimum when min pointer , if the min is popped then min would use this pointer to 
point to the next minimum

class MinStack {
    StackNode current; // pointer for current element on the top of the stack
    StackNode min; // pointer for min

    public MinStack() {
        
        
    }
    
    public void push(int val) {

        StackNode node = new StackNode(val);
        // Stack has no elements
        if(current == null){
            current = node;
            min = node;
            return;
        }

        // Stack has elements

        // updating the current pointer
        StackNode prev = current;
        current = node;
        node.next = prev;

        // checking for min value
        if(node.val < min.val){
            StackNode prevMin = min;
            min = node; // update the min pointer to latest min
            node.nextMin = prevMin; // store the pointer to prev min
        }

    }
    
    public void pop() {
        // removing element from top
        StackNode removedNode = current;
        // updating the min to next min after this element pops
        if(min == removedNode){
            min = removedNode.nextMin;
        }
        current = current.next;
    }
    
    public int top() {
        return current.val;
    }
    
    public int getMin() {
        
         return min.val;
    }
}
// stack node for storing next element as well as next min
class StackNode{
    public int val;
    public StackNode next;
    public StackNode nextMin;

    public StackNode(int value){
        val=value;
    }

}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */